### PR TITLE
feat: add support for other platforms than linux and improved, reading, sending

### PR DESCRIPTION
**Pull request by DjDeveloperr**
towards #1 

- Added `getIPCPath` which returns IPC-pipe path based on OS.
- Added `findIPC` which returns path that is currently active.
- Added reworked packet reading.

Had a few questions before I could start working on support for more commands, for which I've made `NOTE` comments.

Pretty much WIP, haven't tested as I use Windows and there's no unix socket support.. I'll try on a VM.

There's cool things like editing voice setting, creating game lobbies, selecting text channels, etc. which discord-ipc can do 😉

### CODE REVIEW

File: mod.ts [hunk 1]
```diff
@@ -43,15 +39,53 @@ interface Activity {
...
+  }
+}
+
 export async function createClient(): Promise<RichPresence> {
+  const path = await findIPC();
   const client = Object.create(RichPresence.prototype);
+
+  // NOTE(DjDeveloperr): Should we remove this condition since OS-specific logic is handled
+  // in the findIPCPath function? This would maybe just start working once unix sockets are
+  // enabled on Windows.
```
Comment (littledivy) [inline mod.ts:82]: Let's remove the condition and throw for lack of windows support in `findIPCPath` instead.

File: mod.ts [hunk 2]
```diff
@@ -65,6 +99,9 @@ export async function createClient(): Promise<RichPresence> {
   return client;
 }
 
+// NOTE(DjDeveloperr): We should probably use a better name for this,
```
Comment (littledivy) [inline mod.ts:102]: Yeah let's rename this to `DiscordIPC`

File: mod.ts [hunk 3]
```diff
@@ -73,22 +110,42 @@ export class RichPresence {
...
-    await this[_ipcHandle]?.write(encode(0, payload));
-    await read(this[_ipcHandle]);
+    await this.send(OpCode.HANDSHAKE, { v: "1", client_id });
+    await this.#read();
+  }
+
+  #header = new Uint8Array(8);
+  #headerView = new DataView(this.#header.buffer);
+
+  // NOTE(DjDeveloperr): We should keep reading in a loop until it's closed since Discord IPC
```
Comment (littledivy) [inline mod.ts:120]: agreed this can be `TODO`

**Review by littledivy — CHANGES_REQUESTED** [review]

File: mod.ts (hunk 1 above)
Comment (DjDeveloperr) [reply mod.ts]: Done, moved this condition to `findIPCPath`

**Comment by DjDeveloperr** [comment]
Should I be using unique Symbols instead of #private fields (like `_ipcHandle`)?

**Comment by littledivy** [comment]
> Should I be using unique Symbols instead of #private fields (like `_ipcHandle`)?

private fields are fine. `_ipcHandle` is a symbol because I didn't want users to use `DiscordIPC`'s constructor directly

**Review by littledivy — APPROVED** [review]
LGTM Lemme know if its ready to merge

**Comment by DjDeveloperr** [comment]
Private properties seem to have some problem with the way DiscordIPC object is being created

image

**Comment by littledivy** [comment]
Hmm yeah downside of using private properties. Let's just use Symbols for now - I really don't want users to be able to use the constructor as its not part of the public API

**Comment by DjDeveloperr** [comment]
Changed it to use symbols. However, now whenever I try to `read` from `Conn` am getting this error 🤔 

image

**Comment by DjDeveloperr** [comment]
Tested on Ubuntu VM, seems to work

**Comment by DjDeveloperr** [comment]
How would *events* work? Should it extend a EventEmitter (std or x/event), EventTarget + on* or a just Async Iterator?

**Comment by littledivy** [comment]
> How should _events_ work? Should it extend a EventEmitter (std or x/event), EventTarget + on* or a just Async Iterator?

Preferably async iterators

**Pull request merged**